open_instruct/grpo_fast.py

```
@@ -1826,9 +1826,25 @@ def sync_weights_and_prepare_prompts(
     return queries_next, ground_truths_next, datasets_next, dataset_indices
 
 
-def load_data_from_packing_thread(packed_sequences_Q: Queue, num_total_tokens: int):
+def load_data_from_packing_thread(
+    packed_sequences_Q: Queue, num_total_tokens: int, stop_generate_event: threading.Event
+):
     """Get the packed sequences with advantages from the packing thread."""
     with Timer("[Main Thread] 📦 Getting packed sequences from thread"):
+        while not stop_generate_event.is_set():
+            try:
+                packed_data = packed_sequences_Q.get(timeout=2.0)
+                if packed_data is not None:
+                    # got data continue
+                    break
+            except Empty:
+                # keep polling for data
+                continue
+
+        if stop_generate_event.is_set():
+            logger.error("[Main Thread] generate thread has died, cancelling")
+            raise Exception("Generate thread has an exception and died, see error above")
```

`stop_generate_event` is only set when we want to stop the generate thread normally, so I don't think we should raise an exception.

What do you think about wrapping the whole function in the while loop here? 

Something like:

```
with Timer("[Main Thread] 📦 Getting packed sequences from thread"):
    while not stop_generate_event.is_set():
        try:
            packed_data = packed_sequences_Q.get(timeout=2.0)
        except Empty:
            # Keep waiting for more data unless stop_event is set.
            continue
        data_thread_metrics = packed_data["metrics"]
        B = packed_data["B"]
        collated_data = packed_data["collated_data"]
        num_total_tokens += packed_data["num_new_tokens"]
        if B == 0:
            logger.warning("[Main Thread] 🤡 After packing, there is not enough data to train")
            return None, data_thread_metrics, num_total_tokens
        return collated_data, data_thread_metrics, num_total_tokens
```

---

Implemented and added a return None. None None when stop_event.is_set breaks the while loop